astropy/modeling/tests/test_quantities_evaluation.py

```diff
@@ -186,3 +186,281 @@ def test_return_units_scalar(self):
+
+    class ScaleDegrees(Scale):
+        input_units = {'x': u.deg}
+
+    s1 = ScaleDegrees(2)
+    s1.input_units_allow_dimensionless = True
+    s2 = Scale(2)
+
+    cs = s1 | s2
+    cs = cs.rename('TestModel')
+    out = cs(10)
+    assert_quantity_allclose(out, 40 * u.one)
+
+    out = cs(10 * u.arcsec)
+    assert_quantity_allclose(out, 40 * u.arcsec)
+
+    with pytest.raises(UnitsError) as exc:
+        out = cs(10 * u.m)
+    assert exc.value.args[0] == "TestModel: Units of input 'x', m (length), could not be converted to required input units of deg (angle)"
+
+    s1.input_units_allow_dimensionless = False
+
+    cs = s1 | s2
+    cs = cs.rename('TestModel')
+
+    with pytest.raises(UnitsError) as exc:
+        out = cs(10)
+    assert exc.value.args[0] == "TestModel: Units of input 'x', (dimensionless), could not be converted to required input units of deg (angle)"
+
+    s1.input_units_allow_dimensionless = True
+
+    cs = s2 | s1
+    cs = cs.rename('TestModel')
+
+    out = cs(10)
+    assert_quantity_allclose(out, 40 * u.one)
+
+    out = cs(10 * u.arcsec)
+    assert_quantity_allclose(out, 40 * u.arcsec)
+
+    with pytest.raises(UnitsError) as exc:
+        out = cs(10 * u.m)
+    assert exc.value.args[0] == "ScaleDegrees: Units of input 'x', m (length), could not be converted to required input units of deg (angle)"
+
+    s1.input_units_allow_dimensionless = False
+
+    cs = s2 | s1
+
+    with pytest.raises(UnitsError) as exc:
+        out = cs(10)
+    assert exc.value.args[0] == "ScaleDegrees: Units of input 'x', (dimensionless), could not be converted to required input units of deg (angle)"
+
+    s1.input_units_allow_dimensionless = True
+
+    s1 = ScaleDegrees(2)
+    s1.input_units_allow_dimensionless = True
+    s2 = ScaleDegrees(2)
+    s2.input_units_allow_dimensionless = False
+
+    cs = s1 & s2
+    cs = cs.rename('TestModel')
+
+    out = cs(10, 10 * u.arcsec)
+    assert_quantity_allclose(out[0], 20 * u.one)
+    assert_quantity_allclose(out[1], 20 * u.arcsec)
+
+    with pytest.raises(UnitsError) as exc:
+        out = cs(10, 10)
+    assert exc.value.args[0] == "TestModel: Units of input 'x1', (dimensionless), could not be converted to required input units of deg (angle)"
+
+
+def test_compound_return_units():
+    """
+    Test that return_units on the first model in the chain is respected for the
+    input to the second.
+    """
+
+    class PassModel(Model):
+
+        inputs = ('x', 'y')
+        outputs = ('x', 'y')
+
+        @property
+        def input_units(self):
+            """ Input units. """
+            return {'x': u.deg, 'y': u.deg}
+
+        @property
+        def return_units(self):
+            """ Output units. """
+            return {'x': u.deg, 'y': u.deg}
+
+        def evaluate(self, x, y):
+            return x.value, y.value
+
+
+    cs = Pix2Sky_TAN() | PassModel()
+
+    assert_quantity_allclose(cs(0*u.deg, 0*u.deg), (0, 90)*u.deg)
+
+
+def test_missing_key_regression():
```

Comment:
This poses an interesting point about what we want to happen with `return_units`